Calendar: Add loading and empty state

david-szabo97 opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description

**Calendar block**
* Add loading state
* Add empty state
* Make sure the calendar isn't rendered when there are no posts. e028f8a is required to make sure we are consistently rendering the Calendar block in the editor and the frontend. `get_calendar()` returns a calendar HTML output when we set the front page to a static page. When we visit the frontpage in this case, `$posts` is populated with the current static page. Which means the calendar will be rendered with empty posts. (https://core.trac.wordpress.org/ticket/12016)

## How has this been tested?
**Loading state**
1. Open post editor
2. Add Calendar block
3. Save post
4. Refresh
5. You should see a spinner
![image](https://user-images.githubusercontent.com/2256104/117153163-01770900-adbb-11eb-860b-130462465ef3.png)

**Empty state**
1. Delete all posts on your site (you can change them to draft too, just make sure none of them are in `publish` status)
2. Open post editor by creating a new post
3. Add Calendar block
4. It should say: "No posts found."
![image](https://user-images.githubusercontent.com/2256104/117153130-f91ece00-adba-11eb-91be-3feb7448df7f.png)

**Avoid rendering the calendar when there are no posts**
1. Delete all posts on your site (you can change them to draft too, just make sure none of them are in `publish` status)
2. Create a new page, add a Calendar block and save it
3. Go to wpadmin -> Settings -> Reading. Change "Your homepage displays" to "Your static page" and choose the page you just created
4. Open the front page (http://localhost:8888), make sure the calendar isn't rendered.

## Screenshots <!-- if applicable -->

See above

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

david-szabo97 committed:
Add placeholder when no posts exist

david-szabo97 committed:
Add wrapper

david-szabo97 committed:
Don't render the calendar when there are no posts

mattwiebe commented on packages/block-library/src/calendar/index.php:
```diff
@@ -13,7 +13,12 @@
-	global $monthnum, $year;
+	global $monthnum, $year, $wpdb;
+
+	$has_posts = $wpdb->get_var( "SELECT 1 as test FROM $wpdb->posts WHERE post_type = 'post' AND post_status = 'publish' LIMIT 1" );
```
This is a pretty lean query but most of the time we want to cache values like this (in an option or a transient) to avoid hammering the DB (we have tons of users on underpowered shared hosts). That would then require invalidating the cache on every `update_post` action. 

This is just what's done, by default, with the `post_count` option, but unfortunately it only works on Multisite setups. (See `update_post_count()`.)

I don't think this is a blocker, but we'd probably wind up needing to do it sooner or later anyway.

david-szabo97 replied:
Thank you for the review! Added caching in 7b26718

jeyip replied:
> Thank you for the review!

Agreed! Thanks @mattwiebe TIL.

mattwiebe reviewed (commented):
✅  loading state works as described
✅  empty state works as described

Left a comment, let me know what you think.

david-szabo97 committed:
Cache has_published_posts

mattwiebe commented on packages/block-library/src/calendar/index.php:
```diff
@@ -13,10 +13,14 @@
 
-	$has_posts = $wpdb->get_var( "SELECT 1 as test FROM $wpdb->posts WHERE post_type = 'post' AND post_status = 'publish' LIMIT 1" );
-	if ( ! $has_posts ) {
+	$has_published_posts = get_option( 'gutenberg_calendar_block_has_published_posts', null );
```
Nice work! I think I would prefer to see this logic inside of the `block_core_calendar_update_has_published_posts` function itself. That way you don't need to know any implementation logic, just `if ( ! block_core_calendar_update_has_published_posts() ) { return; }`

mattwiebe reviewed (changes requested):
Good stuff here! I made one small request and then this should be good to go

jeyip commented:
## Testing
### Requirements
- [x] Calendar still renders and displays on the frontend when there are published posts
- [x] Calendar displays a spinner while loading
- [x] Empty calendar displays placeholder that says "No posts found."
- [x] Empty calendar does not render on frontend

### Browsers
- [x] Edge
- [x] Firefox
- [x] Safari
- [x] Chrome

jeyip commented on packages/block-library/src/calendar/edit.js:
```diff
@@ -25,20 +28,52 @@ const getYearMonth = memoize( ( date ) => {
+			date: _date,
+			hasPostsResolved: postsResolved,
+			hasPosts:
+				postsResolved && Array.isArray( posts ) && posts.length === 1,
```
`NB`: Would you mind explaining when posts _wouldn't_ be an array? Is it only when it's being fetched? 

In that case, could we write something like `postsResolved && posts && posts.length === 1` or `postsResolved &&  posts?.length === 1` instead?

jeyip reviewed (commented):
`NB`: I found a strange edge case.

1. Ensure there are no published posts
2. Add a calendar block to a draft post
3. Publish the post
4. Switch the post back to a draft in the block editor
5. The placeholder for the calendar block says "Block rendered as empty"

In spite of including `status: 'publish'` in your posts query, when we switch the post back to a draft in step 4,  `getEntityRecords` returns a post with `status: 'draft'`. I'm guessing we're not invalidating some cache properly, but I didn't dig too far. I'm not super concerned because this has a bunch of workarounds (replace the block, refresh the page, etc.), but it might be worthwhile to note in an issue. Let me know what you think 👍 

![2021-05-20 17 35 50](https://user-images.githubusercontent.com/5414230/119065224-e6e38900-b991-11eb-8f1b-1b30c6ff938d.gif)

jeyip reviewed (commented):
+1 whenever the remaining comments are resolved

david-szabo97 committed:
Extract logic as function

david-szabo97 committed:
Add comment

david-szabo97 committed:
Improve readibility

mattwiebe reviewed (approved):
This tests well for me and looks good overall.

I think I was in too much of a rush to explain what I meant in consolidating caching logic, so I went ahead and made my own branch to demonstrate. You can see what I meant in c3fb08ab1ef9fd465f4927994ed4501736527321. 

(Note that this is how I would write it in a WP-like way. Also note that the multisite checks make it more complex, but I like not doing the work twice in that environment.)

I also thought it would be nice to show a message to logged-in users for why the calendar was hidden, so I did that too in e32972ae8bfa2d560e94215c4030928613f2d5f9.

david-szabo97 committed:
Optimize has_published_post flag

- Only update the flag when a post is deleted or status is changed to or from publish
- Use multisite option if available
- Only register actions when on single site

david-szabo97 committed:
Be more specific with the placeholder message

david-szabo97 committed:
Show message for logged in user

Thanks @mattwiebe

david-szabo97 committed:
Update comments

Thanks @mattwiebe

david-szabo97 commented:
Thanks @mattwiebe for helping out here!

Added the check for multisite and optimized the actions to make sure we only update the flag when needed.

david-szabo97 committed:
Fix formatting

david-szabo97 committed:
Comment

jeyip commented:
This still tests well ✅

Are we waiting on anything specifically @david-szabo97?

david-szabo97 commented:
Made a huge change after @mattwiebe review, was hoping he would review it again 😄 

I'm rebasing, hopefully E2E test will pass this time.

mattwiebe reviewed (approved):
Nice work here, let's 🚢  it
